R/resample_data.R

```diff
@@ -28,83 +28,135 @@
+  # Handle all the data sanity checks in outer_level so we don't have redundant error
+  # checks further down the recursion.
+  if(outer_level) {
+    # Optional usage of data.table to speed up functionality
+    # Short-circuit on the is.na to only attempt the package load if necessary.
+    if(is.na(use_dt) && requireNamespace("data.table", quietly=T)) {
+      use_dt = 1
+    } else {
+      use_dt = 0
+    }
 
-  if (any(!ID_labels %in% names(data))) {
-    stop(
-      "One or more of the ID labels you provided are not columns in the data frame provided."
-    )
-  }
+    # User didn't provide an N or an ID label, it's clear they just want a regular bootstrap
+    # of N units by row.
+    if (missing(N) & is.null(ID_labels)) {
+      return(bootstrap_single_level(data, dim(data)[1], ID_label=NULL))
+    }
 
-  if(length(N) > 10) {
-    stop(
-      "Multi-level bootstrap with more than 10 levels is not advised."
-    )
+    # No negative or non-numeric Ns
+    # Note: this should be rewritten when we implement the "ALL" option for a level.
+    if (any(!is.numeric(N) | N%%1 | N<=0)) {
+      stop(
+        "All specified Ns must be numeric and at least 1."
+      )
+    }
+
+    # N doesn't match ID labels
+    if (!is.null(ID_labels) & (length(N) != length(ID_labels))) {
+      stop(
+        "If you provide more than one ID_labels to resample data for multilevel data, please provide a vector for N of the same length representing the number to resample at each level."
+      )
+    }
+
+    # ID_labels looking for some columns we don't have
+    if (any(!ID_labels %in% names(data))) {
+      stop(
+        "One or more of the ID labels you provided are not columns in the data frame provided."
+      )
+    }
+
+    # Excessive recursion depth
+    if(length(N) > 10) {
+      stop(
+        "Multi-level bootstrap with more than 10 levels is not advised."
+      )
+    }
   }
 
   # Single level bootstrap with explicit bootstrapping on a particular cluster variable
+  # this is the inner-most recursion
   if(length(N)==1)
   {
     return(bootstrap_single_level(data,
                                   N[1],
                                   ID_label=ID_labels[1]))
-  } else {
-    # Do the current bootstrap level
-    current_boot_values = unique(data[, ID_labels[1]])
-    sampled_boot_values = sample(1:length(current_boot_values), N[1], replace=TRUE)
-    app = 0
-
-    # Iterate over each thing chosen at the current level
-    results_all = lapply(sampled_boot_values, function(i) {
-      new_results = resample_data(
-        data[data[, ID_labels[1]] == i, ],
-        N=N[2:length(N)],
-        ID_labels=ID_labels[2:length(ID_labels)]
-      )
-    })
-    #res = rbindlist(results_all)
+  }
+
+  # OK, if not, we need to recurse
+
+  # Split indices of data frame by the thing we're strapping on
+  split_data_on_boot_id = split(seq_len(dim(data)[1]), data[,ID_labels[1]])
```

Comment:
`data[[ID_labels[1]]])` will be slightly stricter

Reply:
👍